arch/arm64/core/cache.c

```diff
@@ -57,3 +62,65 @@ size_t arch_dcache_line_size_get(void)
 
 	return dcache_line_size;
 }
+
+
+/*
+ * operation for all data cache
+ * ops:  K_CACHE_INVD: invalidate
+ *	 K_CACHE_WB: clean
+ *	 K_CACHE_WB_INVD: clean and invalidate
+ */
+int arch_dcache_all(int op)
+{
+	uint32_t clidr_el1, csselr_el1, ccsidr_el1;
+	uint8_t loc, ctype, cache_level, line_size, way_pos;
+	uint32_t max_ways, max_sets, dc_val, set, way;
+
+	if (op != K_CACHE_INVD && op != K_CACHE_WB && op != K_CACHE_WB_INVD)
+		return -ENOTSUP;
+
+	clidr_el1 = read_sysreg(CLIDR_EL1);
+
+	/* bit 26 ~ 24 is LoC */
+	loc = (clidr_el1 >> 24) & BIT_MASK(3);
+	if (!loc)
+		return 0;
+
+	for (cache_level = 0; cache_level < loc; cache_level++) {
+		ctype = (clidr_el1 >> cache_level * 3) & BIT_MASK(3);
+		/* No data cache, continue */
+		if (ctype < 2)
+			continue;
+		/* select cache level */
+		csselr_el1 = cache_level << 1;
+		write_sysreg(csselr_el1, CSSELR_EL1);
+		isb();
+		ccsidr_el1 = read_sysreg(CCSIDR_EL1);
+		/* BIT[2:0] is Log2(cache line size) -4 */
+		line_size = (ccsidr_el1 & BIT_MASK(3)) + 4;
+		/* BIT[12:3] is ways - 1 */
+		max_ways = (ccsidr_el1 >> 3) & BIT_MASK(10);
+		/* BIT[27:13] is NumSets - 1*/
+		max_sets = (ccsidr_el1 >> 13) & BIT_MASK(15);
```

Comment:
In addition to using comments I think you can add some define for the numerical values of shifts and masks to make the code clearer.

Reply:
will update, thanks.